### ISSUE TITLE
How do you merge properties of master event and exception?

### ISSUE THREAD

**Issue by afterlogic-support** [ical-org/ical.net#370]
Hi!

If I have an .ICS recurring appointment with an exception and the exception does not have some properties (let's say,  DESCRIPTION), this means these properties must inherit from the master appointment. However, in iCal.Net the corresponding properties remain empty in the exception event.

What is the recommended approach in this situation? Should I check every property of the exception event and, if it's empty, assign its value from the master counterpart? Or, is there a better/automatic way of doing this?

Also, how can I distinguish between missing and empty property? For instance, the exception may have empty DESCRIPTION field denoting that this particular exception has no description (contrary to DESCRIPTION being missing at all which means "inherit it from the master").

Thanks!

**Comment by rianjs** [comment]
I would represent that as two different events:

* Original event with an EXDATE
* A "child" copy of the original event with different properties if desired. Often the easiest way of copying is to reserialize and then deserialize it again into a new `Calendar` or `CalendarEvent`.

To create the child, I would copy the original event in its entirety, and then delete everything associated with recurrences. EXDATE, RDATE, RRULEs, etc. You could tuck the original event id into that child copy if you wanted a "pointer" back to the original. If you wanted to change any properties, you could do so.

There is no formal, "parent-child" relationship in ical.net. You have to maintain that relationship in your application.

**Comment by afterlogic-support** [comment]
Thank you for explanations. However, I'm not sure if we're indeed talking about parsing an .ICS, not building it. I have an .ICS which I need to parse with iCal.Net object model, I'm not building anything (at least in my current task).

The master event may or may not have EXDATE, it's out of my control. It's my input. For instance (but can be anything, actually):

	BEGIN:VEVENT
	SUMMARY:Weekly touch base
	DTSTART;TZID=Asia/Baghdad:20171012T170000
	DTEND;TZID=Asia/Baghdad:20171012T173000
	UID:B24B6767-16B1-4A27-993A-7867213DB2DA
	DTSTAMP:20180521T141428Z
	LAST-MODIFIED:20180316T162333Z
	SEQUENCE:2
	DESCRIPTION:desc
	LOCATION:
	END:VEVENT
	BEGIN:VEVENT
	DTSTART;TZID=Mountain Standard Time:20171019T180000
	DTEND;TZID=Mountain Standard Time:20171019T183000
	UID:B24B6767-16B1-4A27-993A-7867213DB2DA
	RECURRENCE-ID;TZID=Mountain Standard Time:20171019T000000
	DTSTAMP:20180521T141431Z
	END:VEVENT

And I want to be able to parse it to iCal.Net object model the way that the exception event has everything the master event has (with only differences where the exception has them).

**Comment by rianjs** [comment]
I'm still not sure what you're asking. What do you want the outcome to be? You can parse the ics the normal way:

```csharp
var master = Calendar.Load<CalendarEvent>(ical);
```

That'll be a List<CalendarEvent> with 2 VEVENTs inside it. There is no relationship between the two events, in spite of the shared UID. What do you want to happen after that?

**Comment by afterlogic-support** [comment]
Actually, they *are* related. The second one is an exception for the first one (recurring event), indicating that one of the occurrences is an exception and occurs at a different time (in this sample). This exception event often has only the minimum set of properties (which differ it from the master event). So it cannot be used on its own because many properties are missing. It can only used in conjunction with its master event. I need to take the master event and combine it with the exception  - this way I get the remaining properties.

**Comment by afterlogic-support** [comment]
OK, I think I worked it out. I actually sync with Outlook and I found that it already behaves the way I need. When I create child exception, the properties I don't change remain inherited from the master event (this solves the first problem).

And to distinguish between a missing property and explicitly removed property (e.g. when the exception must not have DESCRIPTION rather than borrow it from the master) I can check if it's null or empty in ical.NET objects. If it's null, the property is missing (so it will be borrowed from the master). If it's empty string, it was intentionally cleared and I clear it in the exception. This solves the second problem. Not yet sure if this helps with non-string types but I'll find out.
